File: plugin-template/test/testPlugin.py
import sys

try:
    import unittest
except:
    print("***********************************************************")
    print("* WARNING!! Couldn't import python unittesting framework! *")
    print("* No python tests have been executed                      *")
    print("***********************************************************")
    sys.exit(0)

try:
    import numpy as np
except:
    print("***********************************************************")
    print("* WARNING!! Couldn't import numpy framework!              *")
    print("* No python tests have been executed                      *")
    print("***********************************************************")
    sys.exit(-1)

try:
    import crpropa
except Exception as e:
    print("*** CRPropa import failed")
    print(type(e), str(e))
    sys.exit(-2)
    
try:
    import myPlugin
except Exception as e:
    print("*** myPlugin import failed")
    print(type(e), str(e))
    sys.exit(-2)

class myPluginTests(unittest.TestCase):
	def testSimpleSimulation(self):
		ml = crpropa.ModuleList()

		ml.add(crpropa.SimplePropagation(1 * crpropa.parsec, 100 * crpropa.parsec))
		ml.add(crpropa.MaximumTrajectoryLength(1000 * crpropa.parsec))
		ml.add(myPlugin.MyModule())

		# Preparing source
		source = crpropa.Source()
		source.add(myPlugin.AddMyProperty())

		# Starting Simulation
		ml.run(source, 1)
		# Done


if __name__ == '__main__':
    unittest.main()

File: plugin-template/test/testPlugin.cpp
#include "CRPropa.h"
#include "myPlugin.h"

#include "gtest/gtest.h"



// With this expression we do not need to write `crpropa::` for everything out of the crpropa namespace
using namespace crpropa;

// create a test over TEST(ModuleToTest, WhatToTest){}
TEST(MyModule, SimpleTest) {

	ModuleList sim;
	sim.add(new SimplePropagation(1 * pc, 1 * pc));
	sim.add(new MaximumTrajectoryLength(1000 * pc));
	sim.add(new myPlugin::MyModule());

	// define source to test this module and get candidate with correct flag out of there
	Source source;

	// `AddMyProperty` is added in plugin.h but under the crpropa namespace:
	source.add(new AddMyProperty());
	
	// check if the candidate has the correct property set by the source
	ref_ptr<Candidate> cand = source.getCandidate();
	EXPECT_EQ(cand->getProperty("counter"), uint32_t(0));

	sim.run(cand);

	// in a test you can test your values like this:
	EXPECT_EQ(cand->getProperty("counter"), uint32_t(1000));
	// but there are many more such EXPECT expressions
}

int main(int argc, char **argv) {
	// this runs all above defined tests
	::testing::InitGoogleTest(&argc, argv);
	return RUN_ALL_TESTS();
}